qa/L0_logging/log_format_test.py

```diff
@@ -0,0 +1,271 @@
+            parsed_rows.append(row_data)
+   
+    for row in parsed_rows:
+        assert len(row)==len(header)
+    
+@validator
+def validate_message(message):
+    heading, obj = message.split('\n',1)
+    if heading:
+        try:
+            json.loads(heading)
+        except json.JSONDecodeError as e:
+            raise Exception(f"{e} First line of message in log record is not a valid JSON string")
+        except Exception as e:
+            raise type(e)(f"{e} First line of message in log record is not a valid JSON string")
+    if len(obj):
+        obj = obj.strip()       
+        match = table_regex.search(obj)
+        if match:
+            validate_table(match)
+        else:
+            google.protobuf.text_format.Parse(obj,grpcclient.model_config_pb2.ModelConfig())
+
+class TestLogFormat:
+    @pytest.fixture(autouse=True)
+    def setup(self, request):
+        test_case_name = request.node.name
+        self._server_options = {}
+        self._server_options["log-verbose"] = 256
+        self._server_options["log-info"] = 1
+        self._server_options["log-error"] = 1
+        self._server_options["log-warning"] = 1
+        self._server_options["log-format"] = "default"
+        self._server_options["model-repository"] = os.path.abspath(
+            os.path.join(module_directory, "log_models")
+        )
+        self._server_process = None
+        self._server_options["log-file"] = os.path.join(
+            test_logs_directory, test_case_name + ".server.log"
+        )
+
+    def _launch_server(self, unescaped=None):
+        cmd = ["tritonserver"]
+
+        for key, value in self._server_options.items():
+            cmd.append(f"--{key}={value}")
+
+        env = os.environ.copy()
+
+        if unescaped:
+            env["TRITON_SERVER_ESCAPE_LOG_MESSSAGES"] = "FALSE"
+        elif unescaped is not None:
+            env["TRITON_SERVER_ESCAPE_LOG_MESSSAGES"] = "TRUE"
+
+        self._server_process = subprocess.Popen(
+            cmd,
+            env=env,
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
+        )
+
+        wait_time = 5
+
+        while wait_time and not os.path.exists(self._server_options["log-file"]):
+            time.sleep(1)
+            wait_time -= 1
+
+        if not os.path.exists(self._server_options["log-file"]):
+            raise Exception("Log not found")
+
+    def validate_log_record(self, record, format_regex):
+        match = format_regex.search(record)
+        if match:
+            for field, value in match.groupdict().items():
+                if field in validators:
+                    try:
+                        validators[field](value)
+                    except Exception as e:
+                        raise type(e)(f"{e}\nInvalid {field}: '{match.group(field)}' in log record '{record}'")
+
+        else:
+            raise Exception("Invalid log line")
```

Comment:
Can we invert some checks to avoid the deep nesting here? Something like this?
```python
    def validate_log_record(self, record, format_regex):
        match = format_regex.search(record)
        if not match:
            raise Exception("Invalid log line")

        for field, value in match.groupdict().items():
            if field not in validators:
                continue

            try:
                validators[field](value)
            except Exception as e:
                raise type(e)(f"{e}\nInvalid {field}: '{match.group(field)}' in log record '{record}'")
```